File: Go Solutions/days/day2.go
package days

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func Day2() {
	var fileName string
	if os.Getenv("MODE") == "TEST" {
		fileName = "../inputfiles/Day2Sample.txt"
	} else {
		fileName = "../inputfiles/Day2.txt"
	}
	f, _ := os.Open(fileName)
	defer f.Close()
	var reports [][]int64
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		levels := strings.Split(scanner.Text(), " ")
		var levelInts []int64
		for _, level := range levels {
			levelInt, _ := strconv.ParseInt(level, 0, 0)
			levelInts = append(levelInts, levelInt)
		}
		reports = append(reports, levelInts)
	}
	var safeReports int
	for _, report := range reports {
		if checkReportSafety(report) {
			safeReports++
		}
	}
	fmt.Printf("Part 1 Answer: %d\n", safeReports)
	safeReports = 0
	for _, report := range reports {
		for i := 0; i < len(report); i++ {
			newReport := slices.Clone(report[0:i])
			if i != len(report)-1 {
				newReport = append(newReport, report[i+1:]...)
			}
			if checkReportSafety(newReport) {
				safeReports++
				break
			}
		}
	}
	fmt.Printf("Part 2 Answer: %d\n", safeReports)
}

func checkReportSafety(report []int64) bool {
	if slices.IsSorted(report) {
		return checkDifferenceSafety(report)
	} else {
		slices.Reverse(report)
		if slices.IsSorted(report) {
			return checkDifferenceSafety(report)
		} else {
			return false
		}
	}
}

func checkDifferenceSafety(report []int64) bool {
	for i := 0; i < len(report)-1; i++ {
		if report[i] == report[i+1] || report[i+1]-report[i] > 3 {
			return false
		}
	}
	return true
}

File: Java Solutions/src/Day2.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

/**
 *
 * @author dhc10
 */
public class Day2 {
	
    public static void main(String args[]) {
        Integer safeReports = 0;
        ArrayList<ArrayList<Integer>> reports = new ArrayList<ArrayList<Integer>>();
         try {
            File lists = new File("../inputFiles/Day2.txt");
            Scanner myReader = new Scanner(lists);
            while(myReader.hasNextLine()) {
                String[] levelStrings = myReader.nextLine().split(" ");
                ArrayList<Integer> levels = new ArrayList<Integer>();
                for (int i=0;i<levelStrings.length;i++) {
                    levels.add(Integer.parseInt(levelStrings[i]));
                }
                reports.add(levels);
            }
            myReader.close();
        } catch (FileNotFoundException e) {
            System.out.println("No File Found");
            e.printStackTrace();
        } 
        for (ArrayList<Integer> report : reports) {
            if (isListSorted(report)){
                if (checkReportSafety(report)) {
                    safeReports++;
                }
            }
        }
        System.out.printf("Part 1 Answer: %d\n", safeReports);
        safeReports = 0;
        for (ArrayList<Integer> report : reports) {
            for (int i=0;i<report.size();i++) {
                ArrayList<Integer> newReport = new ArrayList<Integer>(report);
                newReport.remove(i);
                if (isListSorted(newReport)) {
                    if (checkReportSafety(newReport)) {
                        safeReports++;
                        break;
                    } 
                }
            }
        }
        System.out.printf("Part 2 Answer: %d\n", safeReports);
    }

    private static boolean checkReportSafety(ArrayList<Integer> report) {
        for (int i=0;i<report.size()-1;i++) {
            if (Math.abs(report.get(i)-report.get(i+1)) > 3 || report.get(i)==report.get(i+1)) {
                return false;
            }
        }
        return true;
    }

    private static boolean isListSorted(ArrayList<Integer> report) {
        Boolean listSorted = true;
        for (int i = 0;i<report.size()-1;i++) {
            if (report.get(i) > report.get(i+1)) {
                listSorted = false;
                break;
            }
        }
        if (!listSorted) {
            listSorted = true;
            for (int i=0;i<report.size()-1;i++) {
                if (report.get(i)<report.get(i+1)) {
                    listSorted = false;
                    break;
                }
            }
        }
        return listSorted;
    }
}